index/scorch/snapshot_rollback.go

```diff
@@ -15,81 +15,142 @@
-// of a store to previous points in time. A nil return value indicates
-// that no previous snapshots are available.
-func (s *Scorch) PreviousPersistedSnapshot(is *IndexSnapshot) (*IndexSnapshot, error) {
+type RollbackPoint struct {
+	epoch uint64
+	meta  map[string][]byte
+}
+
+func (r *RollbackPoint) GetInternal(key []byte) []byte {
+	return r.meta[string(key)]
+}
+
+// RollbackPoints returns an array of rollback points available
+// for the application to make a decision on where to rollback
+// to. A nil return value indicates that there are no available
+// rollback points.
+func (s *Scorch) RollbackPoints() ([]*RollbackPoint, error) {
 	if s.rootBolt == nil {
-		return nil, nil
+		return nil, fmt.Errorf("RollbackPoints: root is nil")
 	}
 
-	// start a read-only transaction
+	// start a read-only bolt transaction
 	tx, err := s.rootBolt.Begin(false)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("RollbackPoints: failed to start" +
+			" read-only transaction")
 	}
 
-	// Read-only bolt transactions to be rolled back.
+	// read-only bolt transactions to be rolled back
 	defer func() {
 		_ = tx.Rollback()
 	}()
 
 	snapshots := tx.Bucket(boltSnapshotsBucket)
 	if snapshots == nil {
-		return nil, nil
+		return nil, fmt.Errorf("RollbackPoints: no snapshots available")
 	}
 
-	pos := []byte(nil)
+	rollbackPoints := []*RollbackPoint{}
 
-	if is != nil {
-		pos = segment.EncodeUvarintAscending(nil, is.epoch)
-	}
-
-	c := snapshots.Cursor()
-	for k, _ := c.Last(); k != nil; k, _ = c.Prev() {
-		if pos == nil || bytes.Compare(k, pos) < 0 {
-			_, snapshotEpoch, err := segment.DecodeUvarintAscending(k)
-			if err != nil {
-				log.Printf("PreviousPersistedSnapshot:"+
-					" unable to parse segment epoch %x, continuing", k)
-				continue
-			}
+	c1 := snapshots.Cursor()
+	for k, _ := c1.Last(); k != nil; k, _ = c1.Prev() {
+		_, snapshotEpoch, err := segment.DecodeUvarintAscending(k)
+		if err != nil {
+			log.Printf("RollbackPoints:"+
+				" unable to parse segment epoch %x, continuing", k)
+			continue
+		}
 
-			snapshot := snapshots.Bucket(k)
-			if snapshot == nil {
-				log.Printf("PreviousPersistedSnapshot:"+
-					" snapshot key, but bucket missing %x, continuing", k)
-				continue
-			}
+		snapshot := snapshots.Bucket(k)
+		if snapshot == nil {
+			log.Printf("RollbackPoints:"+
+				" snapshot key, but bucket missing %x, continuing", k)
+			continue
+		}
 
-			indexSnapshot, err := s.loadSnapshot(snapshot)
-			if err != nil {
-				log.Printf("PreviousPersistedSnapshot:"+
-					" unable to load snapshot, %v, continuing", err)
-				continue
+		meta := map[string][]byte{}
+		c2 := snapshot.Cursor()
+		for j, _ := c2.First(); j != nil; j, _ = c2.Next() {
+			if j[0] == boltInternalKey[0] {
+				internalBucket := snapshot.Bucket(j)
+				err = internalBucket.ForEach(func(key []byte, val []byte) error {
+					copiedVal := append([]byte(nil), val...)
+					meta[string(key)] = copiedVal
```

Comment:
not understanding the reason here , can't we assign the val directly to the meta map?
may I know, why do we need this step "copiedVal := append([]byte(nil), val...)" ?

Reply:
Values returned by BoltDB are only guaranteed to be valid for the duration of the transaction.